src/type/scalars.js

```diff
@@ -117,24 +117,36 @@ export const GraphQLFloat = new GraphQLScalarType({
   },
 });
 
-function serializeString(value: mixed): string {
-  // Support serializing objects with custom valueOf() functions - a common way
-  // to represent an complex value which can be represented as a string
-  // (ex: MongoDB id objects).
-  const result =
-    value && typeof value.valueOf === 'function' ? value.valueOf() : value;
+// Support serializing objects with custom toJSON() or valueOf() functions -
+// a common way to represent a complex value which can be represented as
+// a string (ex: MongoDB id objects).
+function serializeObject(value: mixed): mixed {
+  if (typeof value === 'object' && value !== null) {
+    if (typeof value.toJSON === 'function') {
```

Comment:
This change is actually breaking - there may be objects which define *both* `valueOf` and `toJSON` which provide different values. If those objects previously worked correctly, this PR would cause them to work incorrectly.

If the goal is to also support objects which define `toJSON` but *do not* define `valueOf` then this could be easily solved by just moving this new addition to *below* the existing check for `valueOf`

Reply:
I did that because: 
```
> typeof ({}).valueOf
'function'
```
But I agree it breaking so I updated this PR to have this instead:
```js
    if (typeof value.valueOf === 'function') {
      const valueOfResult = value.valueOf();
      if (typeof valueOfResult !== 'object') {
        return valueOfResult;
      }
    }
    if (typeof value.toJSON === 'function') {
      return value.toJSON();
    }
```
I also changed tests to explicitly test precidence.
Can you please review new code and tests?